### ISSUE TITLE
Mandate `Access-Control-Allow-Origin: *` for files available in public to allow browser based clients

### ISSUE THREAD

**Issue by bernhardreiter** [oasis-tcs/csaf#653]
If a server is serving CSAF 2.0 related files to the public it MUST set the HTTP-Header `Access-Control-Allow-Origin: *` to allow web browser based clients.

This header setting allows "cross origin resource-sharing" (CORS), so that an application within a browser from a third party domain can access the CSAF information (like security.txt, provider-medata.json, ROLIE files and csaf documents itself).

Without this header for example a single page application cannot access the contents of the files of a CSAF Provider. See the following code example, which you can run in a web browser with javascript enabled (save it as `testcors.html` and open it with the browser).

```html
<!DOCTYPE html>
<html lang="en">
<body>
  <h1>JSON Data Fetched From URL</h1>
  <div id="data"></div>

  <script type="text/javascript">
    // Replace this URL with the actual URL of the JSON file you want to fetch
    var url = 'https://wid.cert-bund.de/.well-known/csaf/provider-metadata.json';
    //var url = 'https://jsonplaceholder.typicode.com/todos/1';

    async function getData() {
        try {
            const response = await fetch(url);
            // variant showing that we indeed get no response with `no-cors`
            //const response = await fetch(url, { mode: 'no-cors' });
            const data = await response.json();

            let ul = document.createElement("ul");
            for (let key in data) {
                let li = document.createElement('li');
                li.textContent = `${key}: ${data[key]}`;
                ul.appendChild(li);
            }
            document.getElementById('data').appendChild(ul);
        } catch (error) {
            console.log(`Error fetching data: ${error}`);
        }
    }

    getData();
  </script>
</body>
</html>
```

You will not see any contents. If you comment out the WID url and use the one from a testing json file, it can be seen that the contents is displayed. A look into the development tools and the network requests shows that a respective header is given.

References:
 * https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
 * https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

**Comment by sthagen** [comment]
During the 2023-09-27 meeting of the TC the members considered the inclusion of the proposed addition to the CSAF v2.1 standard.
> If a server is **serving CSAF 2.0 related files to the public** it MUST set the HTTP-Header `Access-Control-Allow-Origin: *` to allow web browser based clients. 

The TC decided to not include the proposal in CSAF v2.1, instead label it as v2.x and maybe come back to the proposal later.

**Comment by bernhardreiter** [comment]
Even if not included in CSAF 2.1, a recommendation could be mentioned in the FAQ either or CDNs or in a seperate entry.

**Comment by bernhardreiter** [comment]
Raised on csaf-comment today as https://groups.oasis-open.org/discussion/suggestion-require-access-control-allow-origin

**Comment by bernhardreiter** [comment]
## how to configure an ACAO header
### Apache
* ensure to enable the [mod_headers](https://httpd.apache.org/docs/trunk/mod/mod_headers.html) module.
   This depends on how Apache was installed, e.g on Ubuntu and Debian `a2enmod headers` and a restart of httpd
   is most likely the recommended way.
* Add the configuration either to the server configuration file or with `.htaccess` files, the instructions here: https://www.forge12.com/en/blog/enable-cors-in-apache-and-nginx/  look good.
### nginx, freenginx
* Use `add_header` from the [ngx_http_headers_module](https://freenginx.org/en/docs/http/ngx_http_headers_module.html).
* https://www.forge12.com/en/blog/enable-cors-in-apache-and-nginx/ also seems good here for the variants.

**Comment by tschmidtb51** [comment]
@bernhardreiter Do we need to specify the `Access-Control-Allow-Methods` as well?

**Comment by tschmidtb51** [comment]
- [ ] Check preflight
- [ ] Check whether `Access-Control-Allow-Methods` is needed

**Comment by bernhardreiter** [comment]
> Do we need to specify the `Access-Control-Allow-Methods` as well?

No - as the _safe_ methods are allowed by default and those are the ones that are enough for CSAF Providers where public information should be very reachable.

Reference: https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Access-Control-Allow-Methods

> GET, HEAD, and POST are always allowed, regardless of whether they are specified in this header, as they are defined as [CORS-safelisted method](https://fetch.spec.whatwg.org/#cors-safelisted-method)

**Comment by tschmidtb51** [comment]
> > Do we need to specify the `Access-Control-Allow-Methods` as well?
> 
> No - as the _safe_ methods are allowed by default and those are the ones that are enough for CSAF Providers where public information should be very reachable.

Agreed. But aren't browser generating preflight requests automatically?

> A CORS preflight request is a [CORS](https://developer.mozilla.org/en-US/docs/Glossary/CORS) request that checks to see if the CORS protocol is understood and a server is aware using specific methods and headers.
> 
> It is an [OPTIONS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods/OPTIONS) request, using two or three HTTP request headers: [Access-Control-Request-Method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Access-Control-Request-Method), [Origin](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Origin), and optionally [Access-Control-Request-Headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Access-Control-Request-Headers).
>
> A preflight request is automatically issued by a browser and in normal cases, front-end developers don't need to craft such requests themselves.

*Source*: https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request

**Comment by bernhardreiter** [comment]
To be extra clear, this issue is only about the headers for files that should be available to the public. This means TLP:CLEAR only. (As opposed to protected files, including protected feeds. They are are not subject as serving those would require sending some authorisation information, like credentials.)

https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS#simple_requests
> Some requests don't trigger a CORS preflight. Those are called _simple requests_

Here is what is considered a _simple request_:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS#simple_requests

The example to fetch a JSON content works without a preflight request. This is what should be allowed a simple GET for the json files of a CSAF provider's public information. 

See the following expanded test. Save as `testcors3.html` and open in a browser with javascript:
```html
<!DOCTYPE html>
<html lang="en">
<body>
Open the network analysis of the web developers tools 
and reload to see more.
  <script type="text/javascript">
    async function getData(targetElement, resource, options) {
        try {
            const response = await fetch(resource, options); 
            // variant showing that we indeed get no response with `no-cors`
            //const response = await fetch(url, { mode: 'no-cors' });
            const data = await response.json();

            let ul = document.createElement("ul");
            for (let key in data) {
                let li = document.createElement('li');
                li.textContent = `${key}: ${data[key]}`;
                ul.appendChild(li);
            }
            document.getElementById(targetElement).appendChild(ul);
        } catch (error) {
            console.log(`Error fetching data: ${error}`);
        }
    }

    getData();
  </script>

  <h2>JSON Data1 Fetched From URL</h2>   (CORS error expected)
  <div id="data1"></div>
  <script>
    getData("data1",
      'https://wid.cert-bund.de/.well-known/csaf/provider-metadata.json');
  </script>

  <h2>JSON Data2 Fetched From URL</h2>  (after 2 seconds)
  <div id="data2"></div>
  <script>
    setTimeout(() => {
      getData("data2", 'https://jsonplaceholder.typicode.com/todos/1');
    }, 2000);
  </script>

  <h2>JSON Data3 Fetched From URL</h2>  (after 4 seconds)
  <div id="data3"></div>
  <script>
    setTimeout(() => {
      getData("data3", 'https://jsonplaceholder.typicode.com/todos/1',
        { headers: { "X-PINGOTHER": "pingpong" } });
    }, 4000);
  </script>

</body>
</html>
```

The third example adds an extra header, which triggers the preflight request.
But the second example is a simple request and in the Firefox and Chromium browsers I've tied, data2 works without preflight request.

My conclusion is that we do not need to care about preflight request for the regular use case to just request the public files.
